boomonkeyboo/scratch_project
# Created WebSocket Events

**Issue Type**

**Description**
Set up server-side event listeners and emissions

**Server Event Listeners/Requests from Client**
- "register": (username)
- "joinRoom": (room)
- "leaveRoom": ()
- "message": (message)

**Server Event Emissions/Client Events to Listen For**
- "systemMessage": string, messages from server to client when a client has activity other than a chatroom message
- "message": `{username, message}`, JSON consisting of username and message contents

**Steps to Validate Feature**

0. Run `npm install` to install any updated dependencies
1. Install Postman and test collection in [Postman group](https://lively-meadow-311823.postman.co/workspace/Scratch-Project~a6efac18-d4d6-4119-9aff-e9cbd391d2ea/collection/651f8573cae43373c3ded881?action=share&creator=30175074)
2. Run npm run dev to start back end

Create connections and register usernames
1. Start a new Socket.IO connection to ws://localhost:3001, username will be anonymous and room will be lobby
3. Add event listener for "message" and "systemMessage" and turn both listeners on
4. Send a message with Event name "register" with body "user1"
5. Start a 2nd Socket.IO connection to ws://localhost:3001, username will be anonymous and room will be lobby
6. Add event listener for "message" and "systemMessage" and turn both listeners on
7. Send a message with Event name "register" with body "user2"

Test Messages
1. From user1 send a message with Event name "message" and body "test message from user1"
2. user2 will see user1's message in room "lobby" formatted as {"username: "user1", "message": "test message from user1"}
3. From user1 send a message with Event name "joinRoom" and body "coding"
4. From user1 send a message with Event name "message" and body "test message to coding room only"
5. user2 will not see message since user2 is in lobby
6. From user2 send a message with Event name "joinRoom" and body "coding"
8. From user1 send a message with Event name "message" and body "hello user2"
9. user2 will see an event with the name "message" and contents {"username": "user1", "message": "hello user2"}

**Previous behavior**
Single chat room

**Expected behavior**
- Users will be in a room named "lobby" by default
- Users can join rooms by sending a message with Event name "joinRoom" and body containing a room name
- Users can send messages that will only be viewed by users in the same room as the user is currently in

**Screenshots & Videos** 
user1 registers username
![image](https://github.com/boomonkeyboo/scratch_project/assets/13823341/9d38af1a-0616-4ed0-9e33-06a403d2dd41)

user1 joins sports room
![image](https://github.com/boomonkeyboo/scratch_project/assets/13823341/7601c2d1-186d-4e39-a538-50dcbc058d89)

user1 sends a message to sports room only
![image](https://github.com/boomonkeyboo/scratch_project/assets/13823341/fb5406af-62e1-4d7d-ac54-209a1fd3fe60)

user2 can view the message send by user1 only if user2 is in sports room. user2 cannot see from lobby
![image](https://github.com/boomonkeyboo/scratch_project/assets/13823341/0108eae2-83b5-445d-a1aa-fcea006ddcd5)


**Additional context**  
Testing requires a long sequence in WebSockets. Consider adding Jest or Mocha code for testing later.
